Add public key support

RopoMen opened the pull request:
# Description
- re-factored "crypto.ts" to handle PEM files in more general way.
- Added unit tests for new functions.
- Re-factored internally implementation to use "key info" (or similar) terminology instead of talking only about certificates. Certificate term is ok, if we known that we are handling certificate.

These changes are not making any changes into used interface, only internal re-factoring.

I will make changes to README.md before merge, because it is not currently talking about public key support. Public key support comes through `cert` property, if PEM file is well formatted then it is accepted.

RopoMen commented:
Resolves https://github.com/node-saml/node-saml/issues/206

cjbarth commented:
Thank you for this, but all these code scanning alerts will have to be addressed. If you need help figuring out how to make the regex less catastrophic, please post here and we'll see what help we can provide.

It looks like the new code you added isn't 100% covered by tests. Please have a look and see what can be done about that.

I look forward to seeing your README updates; our README definitely needs a little TLC.

Also, if it is easy to add a test to add coverage to another part of the code that is currently untested, perhaps a branch, please consider that; sometimes I find that when I'm writing a test trying to get complete coverage for my new code, I see that I can make one tweak to a new test and cover a previously untested branch in some other code.

RopoMen commented:
@cjbarth 

1. `..but all these code scanning alerts will have to be addressed.` ofc. I'll check what is the best way to test potential fix before pushing changes. (I have not used Github code scanning tools before)
2. `It looks like the new code you added isn't 100% covered by tests. `, I'll check what improvements there can be made. Those PEM handling functions should be covered as much as possible. My guess is that assertions should be tested better.
3. README.md updates
4. Add at least one unit test for saml.ts where well formatted PEM public key is given into library.
5. Check that is there need to add tests for saml.ts `getKeyInfosAsPem()` function. (most likely yes)

cjbarth commented:
> 1. `..but all these code scanning alerts will have to be addressed.` ofc. I'll check what is the best way to test potential fix before pushing changes. (I have not used Github code scanning tools before)

One of the best ways is to just look at the manual: https://codeql.github.com/codeql-query-help/javascript/js-polynomial-redos/. As suggested, anchoring the string or using some lookbehind helps a lot.

RopoMen commented:
@cjbarth I have now;
- Added more tests for crypto.ts
- Fixed regex issues
- Added test for public key
- Added tests for resolveAndParseKeyInfosToPem()
- Updated readme to contain mention about public key support

cjbarth commented on src/crypto.ts:
```diff
@@ -1,48 +1,93 @@
+export const stripPemHeaderAndFooter = (certificate: string): string => {
+  return certificate
+    .replace(/(\r\n|\r)/g, "\n")
+    .replace(/-----BEGIN [A-Z\x20]{1,48}-----\n?/, "")
```
Why do you use `\x20` instead of a space?

cjbarth commented on src/saml.ts:
```diff
@@ -40,14 +40,38 @@ import {
+      .then((certs) => {
+        assertRequired(certs, "callback didn't return cert");
+
+        keyInfosToHandle.push(certs);
```
You might have to use a spread operator here as we don't want to have nested arrays.

cjbarth commented on test/crypto.spec.ts:
```diff
@@ -43,22 +96,144 @@ describe("crypto.ts", function () {
+      });
+
+      it("should throw with false", function () {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Instead of disabling, let's use `@ts-expect-error`.

cjbarth replied:
We don't want to disable checking, we want to inform the linter that an error should be here.

cjbarth commented on tsconfig.json:
```diff
@@ -4,9 +4,9 @@
     /* Basic Options */
     // "incremental": true,                   /* Enable incremental compilation */
-    "target": "es2018" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */,
+    "target": "es2019" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */,
```
What is the reason for this change? I'm not against it, but it does seem like something that we should save for a semver-major change.

cjbarth commented on src/crypto.ts:
```diff
@@ -1,48 +1,93 @@
+const BASE64_REGEX =
+  /^(?:[A-Za-z0-9\+\/]{4})*(?:[A-Za-z0-9\+\/]{2}==|[A-Za-z0-9\+\/]{3}=|[A-Za-z0-9\+\/]{4})$/m; // eslint-disable-line no-useless-escape
+
+export const PemLabel = {
```
If we have to export it, then it seems it should be in `types.ts`.

cjbarth commented:
Thanks for all your hard work and the clarity of your code. I see that this patch is almost at 100%, good job! Keep at it!

RopoMen commented:
@cjbarth is there something that should be done to get this PR into master?
- DeepScan finding is false-positive
- codecov finding is minor

cjbarth commented on src/saml.ts:
```diff
@@ -638,30 +668,18 @@ class SAML {
+      return await resolveAndParseKeyInfosToPem(this.options);
+    } else if (this.pemFiles.length > 0) {
+      // Return already cached PEM files.
+      return this.pemFiles;
```
This is part of the new functionality that you've added that is completely uncovered by tests. Can you please add one here to make sure that caching is working as intended?

cjbarth commented on src/pem-label.enum.ts:
```diff
@@ -0,0 +1,7 @@
+export const PemLabel = {
```
What was the reasoning behind putting this type in its own file instead of with all the other types?

cjbarth commented on test/tests.spec.ts:
```diff
@@ -701,7 +701,7 @@ describe("node-saml /", function () {
         const container = { SAMLResponse: base64xml };
         const samlObj = new SAML({
-          cert: "-----BEGIN CERTIFICATE-----" + TEST_CERT + "-----END CERTIFICATE-----",
+          cert: `-----BEGIN CERTIFICATE-----\n${TEST_CERT}\n-----END CERTIFICATE-----`,
```
This technically changes the test. Will the code still work if there isn't a newline after the opening segment or before the closing segment? I know that is non-standard, but, since it is supported that way in a test, we should probably continue supporting it until a semver-major.

RopoMen commented:
@cjbarth Hi, is there something that you are expecting me to work with this PR? Or do you have time allocation issues with this PR? I would like to work to get this PR finished.

cjbarth commented:
It seems that my previous PR comments are unresolved. Why not start there? I'm also interested in addressing those regex warnings, but I need more time to review what can be done there.

RopoMen commented:
DeepScan stuff fixed, I'll check if I can get coverage a bit better.

RopoMen commented:
@cjbarth Hi, I resolved most of the conversations on couple of weeks ago, but there are still those two open conversations. For now I need more feedback from you in order to get things moving.

If there are some issues which needs to be addressed in specific way other than those two open conversations, I wish you point out those explicitly where the issue is and how you want it to be fixed so that I can make those fixes.

cjbarth commented:
I'm sorry for the delay; I was unavailable for a while. I'm back to giving this project more attention. I've read over the changes you've made and they look good. I'll take time to take a second look at my comments that are still open and see if they are still valid or if I need to add more details.

cjbarth commented:
@RopoMen , I've pushed up a lot of changes to improve this code. We don't want to export code just for testing, so many tests were changed, or even removed, for that reason. Many tests were redundant.

I see the point about the `codecov` tool. I'll merge if you agree with the changes I've made. I look forward to your feedback.

RopoMen commented:
> @RopoMen , I've pushed up a lot of changes to improve this code. We don't want to export code just for testing, so many tests were changed, or even removed, for that reason. Many tests were redundant.
> 
> I see the point about the `codecov` tool. I'll merge if you agree with the changes I've made. I look forward to your feedback.

Hi @cjbarth 
You can merge this if you are done. I don't mind about the changes, main point is more important.

Thank you to be active again 😁
